src/hotspot/share/gc/shenandoah/shenandoahEvacOOMHandler.cpp

```diff
@@ -164,6 +192,27 @@ void ShenandoahEvacOOMHandler::unregister_thread(Thread* thr) {
   assert(!ShenandoahThreadLocalData::is_oom_during_evac(thr), "TL oom-during-evac must be turned off");
 }
 
+// The current thread failed to allocate memory required by evacuation.  Perform the following:
+//
+// Upon entry:
+//
+//  1. The current thread is known to be authorized to allocate for evacuation.
+//
+// Upon return:
+//
+//  1. The OOM bit is set for every counter.
+//  2. This thread's thread-local is_oom_during_evac flag is true, denoting that this thread is no longer authorized
+//     to perform evacuation allocations.
+//  3. The count of threads authorized to evacuate for allocation has been decremented, because this thread is no
+//     longer authorized.
+//  4. We have waited for all evacuating threads to stop allocating, after which it is safe for this thread to resolve
```

Comment:
This is where I think the implementation breaks down. If thread `A` is the _first_ thread to attempt evacuation and it fails _before_ any other thread has attempted to evacuate, then thread `A` "believes" that no other threads are evacuating and it returns immediately. How does the protocol then prevent other threads from proceeding with evacuation.

Reply:
Even when thread A is the first thread to attempt evacuation, it will:
1. Iterate through all counters and set the OOM bit on each one.  Any new thread that attempts to enter_evacuation() will check the OOM bit on its respective counter.  If the OOM bit is already set, that new thread will not be authorized to allocate.  If the thread happens to enter before this thread A has set its OOM bit, then we resolve this in the next step.
2. After setting the OOM bit on each counter, we wait_for_no_evac_threads before we consider it safe to make use of a from-space pointer.

Reply:
So what about this scenario?
1. Thread A is setting the OOM bits on all counters, has decremented its own count, but has not yet finished setting all OOM bits
2. Thread B newly tries to enter_evacuation() so it invokes register_thread, and and finds the OOM bit is set.  It waits for no_evac_threads() and this immediately returns, because there are no evac thread running at this moment.
3. Now thread C newly tries to enter_evacuation() so it invokes register thread, but thread A has not yet set this thread's OOM bit, so thread C proceeds to enter_evac with authorization to evacuate.

This does appear to be a bug.  I think the fix is that register_thread also needs to repeat the loop that sets all OOM bits before it waits for no evac threads.

Reply:
I agree. Though I haven't reproduced the scenario you're describing, I stepped through this code a few times with failure scenarios.